### PR TITLE
Initial integrating of Wazuh Manager and Agents into JupiterOne  

### PR DESCRIPTION
**Pull request by lincolnfzr**
This initial integration of Wazhuh captures the Manager and its Agents.  The Manager is also classified as the Account class entity. The calls are made through the RESTful services of Wazuh-API.

Most of the work is done in the provider.ts and converter.ts services.

screen shot 2019-03-03 at 11 20 53 pm

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -53,7 +52,17 @@ with no existing data:
 
 ### Environment Variables
 
-* `GRAPH_DB_ENDPOINT` - `"localhost"`
+- `GRAPH_DB_ENDPOINT` - `"localhost"`
+
+- `WAZUH_API_USER_CONFIG`=`user ID`
+- `WAZUH_API_SECRET_CONFIG`=`pass word`
+- `WAZUH_API_HOST_CONFIG`=`Wazuh API Server Host`:55000
+- `WAZUH_API_PROTOCAL_CONFIG`=`http://`
```
Comment (aiwilliams) [inline README.md]: I think it would be good to [use the term scheme](https://en.wikipedia.org/wiki/URL#Syntax). Also, I'd recommend removing the `://` part and placing that in the URL construction code.

```suggestion
- `WAZUH_API_SCHEME` - `"http"`
```

File: package.json [hunk 2]
```diff
@@ -17,7 +17,9 @@
     "test": "jest"
   },
   "dependencies": {
-    "bunyan": "^1.8.12"
+    "@types/request": "^2.48.1",
```
Comment (aiwilliams) [inline package.json]: This should be in `devDependencies` since it isn't used at runtime nor should it be installed into projects that use the compiled NPM module (transitive dependencies).

File: src/converters.ts [hunk 3]
```diff
@@ -0,0 +1,109 @@
+import {
+  EntityFromIntegration,
+  RelationshipFromIntegration
+} from "@jupiterone/jupiter-managed-integration-sdk";
+import { Agent, WazuhManager } from "./provider";
+
+export const ACCOUNT_ENTITY_TYPE = "provider_account";
```
Comment (aiwilliams) [inline src/converters.ts]: This constant is not used.
```suggestion
```

File: src/converters.ts [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+export const ACCOUNT_ENTITY_CLASS = "Account";
+
+export const WAZUH_MANAGER_ENTITY_TYPE = "provider_wazuhmanager";
```
Comment (aiwilliams) [inline src/converters.ts]:
```suggestion
export const WAZUH_MANAGER_ENTITY_TYPE = "wazuh_manager";
```

File: src/converters.ts [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+export const WAZUH_MANAGER_ENTITY_CLASS = "Wazuhmanager";
```
Comment (aiwilliams) [inline src/converters.ts]: Entities should have a class defined in the [data model](http://docs.jupiterone.io/en/latest/docs/jupiterone-data-model.html). I'd stick with just `Account` for now.

```suggestion
export const WAZUH_MANAGER_ENTITY_CLASS = "Account";
```

File: src/converters.ts [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+
+export const AGENT_ENTITY_TYPE = "provider_agent";
```
Comment (aiwilliams) [inline src/converters.ts]:
```suggestion
export const AGENT_ENTITY_TYPE = "wazuh_host_agent";
```

File: src/converters.ts [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+export const AGENT_ENTITY_CLASS = "Agent";
```
Comment (aiwilliams) [inline src/converters.ts]: The data model specifies a [`HostAgent` class](http://docs.jupiterone.io/en/latest/@lifeomic/jupiter-data-model/docs/HostAgent.html).

```suggestion
export const AGENT_ENTITY_CLASS = "HostAgent";
```

File: src/converters.ts [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+
+export const ACCOUNT_WAZUH_MANAGER_RELATIONSHIP_TYPE =
+  "provider_account_wazuhmanager";
+export const ACCOUNT_WAZUH_MANAGER_RELATIONSHIP_CLASS = "HAS";
```
Comment (aiwilliams) [inline src/converters.ts]: `ACCOUNT_WAZUH_MANAGER_RELATIONSHIP_TYPE` and `ACCOUNT_WAZUH_MANAGER_RELATIONSHIP_CLASS` do not appear to be used.

File: src/converters.ts [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+
+export const WAZUH_MANAGER_AGENT_RELATIONSHIP_TYPE =
+  "provider_wazuhmanager_agent";
```
Comment (aiwilliams) [inline src/converters.ts]:
```suggestion
  "wazuh_manager_agent";
```

File: src/converters.ts [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+}
+
+export function createWazuhManagerEntities(
+  data: WazuhManager
+): WazuhManagerEntity {
+  // Note the manager is also the account entity
+  return {
+    _key: `${WAZUH_MANAGER_ENTITY_TYPE}-${data.id}`,
+    _type: WAZUH_MANAGER_ENTITY_TYPE,
+    _class: [WAZUH_MANAGER_ENTITY_CLASS, ACCOUNT_ENTITY_CLASS],
```
Comment (aiwilliams) [inline src/converters.ts]: As mentioned, `Account` is sufficient. The other class does not fit the data model. Also, if we add another class, `WAZUH_MANAGER_ENTITY_CLASS` should become an `Array` of class values.

```suggestion
    _class: WAZUH_MANAGER_ENTITY_CLASS,
```

File: src/converters.ts [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+}
+
+export interface AgentEntity extends EntityFromIntegration, Agent {
+  ownerId: string;
+}
+
+export function createWazuhManagerEntities(
+  data: WazuhManager
+): WazuhManagerEntity {
+  // Note the manager is also the account entity
```
Comment (aiwilliams) [inline src/converters.ts]:
```suggestion
```

File: src/provider.ts [hunk 12]
```diff
@@ -0,0 +1,180 @@
...
+      .then(data => {
+        return data;
+      })
+      .catch(error => {
+        throw new Error(`${error}: fetching agents`);
+      });
+  }
+
+  private mapToAgent(a: string): Agent {
+    const agent = JSON.parse(JSON.stringify(a));
```
Comment (aiwilliams) [inline src/provider.ts]: Why is this necessary?

File: src/provider.ts [hunk 13]
```diff
@@ -0,0 +1,180 @@
...
+
+export interface ProviderConfig {
+  userId: string;
+  secret: string;
+  baseUrlProtocal: string;
+  baseUrlHost: string;
+}
+
+export class ProviderClient {
+  private manager_Url: string;
```
Comment (aiwilliams) [inline src/provider.ts]:
```suggestion
  private managerUrl: string;
```

File: src/provider.ts [hunk 14]
```diff
@@ -0,0 +1,180 @@
...
+  private agent_Url: string;
```
Comment (aiwilliams) [inline src/provider.ts]:
```suggestion
  private agentUrl: string;
```

File: src/provider.ts [hunk 15]
```diff
@@ -0,0 +1,180 @@
...
+
+  constructor(_providerConfig: ProviderConfig) {
```
Comment (aiwilliams) [inline src/provider.ts]: Our preference is to avoid an `_` prefix notation. Is there something this is meant to communicate?

File: src/provider.test.ts [hunk 16]
```diff
@@ -0,0 +1,30 @@
...
+const integrationConfig: ProviderConfig = {
+  userId: process.env.WAZUH_API_USER_CONFIG,
+  secret: process.env.WAZUH_API_SECRET_CONFIG,
+  baseUrlHost: process.env.WAZUH_API_HOST_CONFIG,
+  baseUrlProtocal: process.env.WAZUH_API_PROTOCAL_CONFIG,
+};
+
+test("Wazuh Manager information and connection", async () => {
+  // const executionContext = createTestIntegrationExecutionContext();
+  let providerClient: ProviderClient = new ProviderClient(integrationConfig);
```
Comment (aiwilliams) [inline src/provider.test.ts]: I'm going to guess that the tests will not run if credentials are required in the CI/CD pipeline. We'll need to support two test targets, `yarn test` and `yarn test:ci`, and move these integration tests into `<rootDir>/test/integration/provider.test.ts`, and there'll need to be a Wazuh server running somewhere that CI/CD machines can reach. So, I think for now, you'd have to use `jest.mock` to stub out provider API responses.

Check out some of the provider tests in https://github.com/JupiterOne/jupiter-integration-onelogin/pull/2 for examples/inspiration.

File: src/executionHandler.ts [hunk 17]
```diff
@@ -2,16 +2,71 @@ import {
...
+    oldManagerAgentRelationships
+  ] = await Promise.all([
+    graph.findEntitiesByType<WazuhManagerEntity>(WAZUH_MANAGER_ENTITY_TYPE),
+    graph.findEntitiesByType<AgentEntity>(AGENT_ENTITY_TYPE),
+    graph.findRelationshipsByType(WAZUH_MANAGER_AGENT_RELATIONSHIP_TYPE)
+  ]);
+
+  const managerEntity: WazuhManagerEntity = await provider
+    .fetchManager()
+    .then(manager => {
```
Comment (aiwilliams) [inline src/executionHandler.ts]: We prefer to avoid using promise chaining since we have the latest JS at our disposal thanks to TypeScript and Babel.

File: src/initializeContext.ts [hunk 18]
```diff
@@ -0,0 +1,32 @@
+import {
+  GraphClient,
+  IntegrationExecutionContext,
+  // EntityFromIntegration,
```
Comment (aiwilliams) [inline src/initializeContext.ts]:
```suggestion
```

File: src/provider.ts [hunk 19]
```diff
@@ -0,0 +1,180 @@
...
+      throw new Error("User ID and Password required");
+    } else {
+      this.manager_Url = `${_providerConfig.baseUrlProtocal}${
+        _providerConfig.userId
+      }:${_providerConfig.secret}@${_providerConfig.baseUrlHost}/manager/info`;
+      this.agent_Url = `${_providerConfig.baseUrlProtocal}${
+        _providerConfig.userId
+      }:${_providerConfig.secret}@${_providerConfig.baseUrlHost}/agents`;
+      console.log(this.manager_Url);
+      console.log(this.agent_Url);
```
Comment (aiwilliams) [inline src/provider.ts]: These `console.log` statements should be removed.

File: src/provider.ts [hunk 20]
```diff
@@ -0,0 +1,180 @@
...
+          compilationDate: provider.data.compilation_date,
+          version: provider.data.version,
+          opensslSupport: provider.data.openssl_support,
+          maxAgents: provider.data.max_agents,
+          rulesetVersion: provider.data.ruleset_version,
+          path: provider.data.path,
+          tzName: provider.data.tz_name,
+          type: provider.data.type,
+          tzOffset: provider.data.tz_offset,
+          id: "0"
```
Comment (aiwilliams) [inline src/provider.ts]: Why is this always `"0"`?

File: README.md [hunk 21]
```diff
@@ -53,7 +52,17 @@ with no existing data:
 
 ### Environment Variables
 
-* `GRAPH_DB_ENDPOINT` - `"localhost"`
+- `GRAPH_DB_ENDPOINT` - `"localhost"`
+
+- `WAZUH_API_USER_CONFIG`=`user ID`
```
[the remark this answers is not in the document]
Comment (lincolnfzr) [reply README.md]: That's correct, there is no default.  These environment variables only need to be set in a standalone/test run.  The JupiterOne integration UI is expected to set them in context.instance.config as type ProviderConfig.

File: src/provider.test.ts (hunk 16 above)
Comment (lincolnfzr) [reply src/provider.test.ts]: Yes.  There is work to be done here. I planned on mocking the fetch functions and create fake data with faker.  No Wazuh server would need to be running to test the code.  It would leverage the JupitorOne docker backend. Thanks for the pointer.

File: src/executionHandler.ts (hunk 17 above)
Comment (lincolnfzr) [reply src/executionHandler.ts]: I will convert all of the chains to async/await

File: src/converters.ts (hunk 8 above)
Comment (lincolnfzr) [reply src/converters.ts]: I intended to create an account --> manager relationship.  I'll delete them.

File: src/provider.ts (hunk 20 above)
Comment (lincolnfzr) [reply src/provider.ts]: The WazuhManager does not have an id, per our conversation.  I am initializing it to zero here.

File: src/provider.ts (hunk 12 above)
Comment (lincolnfzr) [reply src/provider.ts]: I'll clean it up with the other chaining changes.

File: src/converters.ts [hunk 22]
```diff
@@ -0,0 +1,107 @@
...
+  RelationshipFromIntegration
+} from "@jupiterone/jupiter-managed-integration-sdk";
+import { Agent, WazuhManager } from "./provider";
+
+export const ACCOUNT_ENTITY_CLASS = "Account";
+
+export const WAZUH_MANAGER_ENTITY_TYPE = "wazuh_manager";
+export const WAZUH_MANAGER_ENTITY_CLASS = "Account";
+
+export const AGENT_ENTITY_TYPE = "wazuh_host_agent";
```
Comment (trust-no-one) [inline src/converters.ts]: how about we change this type to `wazuh_agent` for simplicity? Or are there other types of agents supported/provided by Wazuh?

File: src/converters.ts [hunk 23]
```diff
@@ -0,0 +1,107 @@
...
+export const AGENT_ENTITY_CLASS = "HostAgent";
+
+export const ACCOUNT_WAZUH_MANAGER_RELATIONSHIP_TYPE =
+  "provider_account_wazuhmanager";
+export const ACCOUNT_WAZUH_MANAGER_RELATIONSHIP_CLASS = "HAS";
+
+export const WAZUH_MANAGER_AGENT_RELATIONSHIP_TYPE =
+  "wazuh_manager_agent";
```
Comment (trust-no-one) [inline src/converters.ts]: we typically include the relationship verb in the type - for example, this one would probably be `wazuh_manager_has_agent`

File: src/converters.ts [hunk 24]
```diff
@@ -0,0 +1,107 @@
+import {
+  EntityFromIntegration,
+  RelationshipFromIntegration
+} from "@jupiterone/jupiter-managed-integration-sdk";
+import { Agent, WazuhManager } from "./provider";
+
+export const ACCOUNT_ENTITY_CLASS = "Account";
+
+export const WAZUH_MANAGER_ENTITY_TYPE = "wazuh_manager";
+export const WAZUH_MANAGER_ENTITY_CLASS = "Account";
```
Comment (trust-no-one) [inline src/converters.ts]: do you consider an Wazuh Manager instance an Account? if so, this makes sense. 

Can you have a Wazuh Account with more than one Manager? If that's the case, we should probably create `wazuh_account` and `wazuh_manager` separately.

Comment (aiwilliams) [reply src/converters.ts]: @trust-no-one, it seems that a Wazuh Manager is the root thing in a Wazuh "account", judging by the fact that there is no way to specify which Manager in the API calls https://documentation.wazuh.com/current/user-manual/api/reference.html#manager.

I think it reasonable to store Manager info on an entity that is `_class: 'Account'`, but it is going to be expected that the `_type` of that entity be something like `wazuh_account`. I think we should consider changing to:

```js
export const WAZUH_ACCOUNT_ENTITY_TYPE = "wazuh_account";
export const WAZUH_ACCOUNT_ENTITY_CLASS = "Account";
```

Then just use the Manager APIs to populate the account entity, OR, perhaps better, if we want the graph `Account -> Manager -> All Stuff`, this would be more consistent with Veracode, which does `Account -> Service -> All Stuff`.

Comment (lincolnfzr) [reply src/converters.ts]: set to [Service, Control] per slack thread.

File: .gitignore [hunk 25]
```diff
@@ -2,6 +2,7 @@ node_modules/
 build/
 coverage/
 dist/
+src/.DS_Store
```
Comment (aiwilliams) [inline .gitignore]: I'd recommend adding `.DS_Store` to `~/.gitignore`, since this will get you as a Mac user in every git repo you open in Finder 😄

File: package.json [hunk 26]
```diff
@@ -14,11 +14,15 @@
     "build:src": "babel src --extensions '.ts' --out-dir dist",
     "build:files": "ts-node tools/copy-dist-files.ts",
     "build:types": "tsc --project tsconfig.dist.json --declaration --emitDeclarationOnly",
+    "build:publish": "yarn build && cd dist && yarn publish --access public",
+    "lint": "yarn lint:ts && yarn lint:md",
+    "lint:md": "remark .",
+    "lint:ts": "tslint --project .",
+    "format": "yarn prettier --write '**/*.{ts,js,json,md}'",
```
Comment (aiwilliams) [inline package.json]: Excellent, thanks so much for updating this to match the example integration @lincolnfzr !

File: src/converters.test.ts [hunk 27]
```diff
@@ -0,0 +1,152 @@
...
+      osVersion: "16.04.5 LTS",
+      osCodename: "Xenial Xerus",
+      osArch: "x86_64",
+      osMinor: "04",
+      id: "002",
+    },
+  ];
+}
+
+test("Wazuh Manager being converted to Manager Entity where manager = manager entyity.", () => {
```
Comment (aiwilliams) [inline src/converters.test.ts]: Not a biggie at all, but `entyity` is typo.

File: src/converters.test.ts [hunk 28]
```diff
@@ -0,0 +1,152 @@
...
+  const m: WazuhManager = fetchManager();
+  const managerEntity: WazuhManagerEntity = createWazuhManagerEntities(m);
+
+  expect(m.id).toEqual(managerEntity.id);
```
Comment (aiwilliams) [inline src/converters.test.ts]: `expect` is meant to receive the test subject, with `toEqual` receiving the expected value. Something like, "expect the value generated by function y to equal value x".

File: src/executionHandler.ts [hunk 29]
```diff
@@ -12,61 +12,53 @@ import {
...
+        [
+          ...persister.processRelationships(
+            oldManagerAgentRelationships,
+            createWazuhManagerAgentRelationships(managerEntity, agentEntities),
+          ),
+        ],
+      ]),
+    };
+  } catch (error) {
+    throw new Error(error);
```
Comment (aiwilliams) [inline src/executionHandler.ts]: What is the reason for wrapping the `error`?

File: src/provider.test.ts (hunk 16 above)
Comment (lincolnfzr) [reply src/provider.test.ts]: I mocked the manager and agent when WAZUH_API_INTEGRATION  is not set or is set to '0', the default.

File: src/initializeContext.test.ts [hunk 30]
```diff
@@ -1,8 +1,34 @@
...
 
-test("creates provider client", () => {
+test("creating Wazuh client", async () => {
   const executionContext = createTestIntegrationExecutionContext();
+
+  const integrationConfig: ProviderConfig = {
+    userId:
+      process.env.WAZUH_API_USER !== undefined
+        ? process.env.WAZUH_API_USER
+        : "foo",
```
Comment (aiwilliams) [inline src/initializeContext.test.ts]: These can be written as:

```js
userId: process.env.WAZUH_API_USER || "foo"
```

That said, why would the tests want to use the environment values? Is the goal to allow running these tests against the actual APIs?

File: .gitignore (hunk 25 above)
Comment (lincolnfzr) [reply .gitignore]: Done thanks

File: src/invocationValidator.ts [hunk 31]
```diff
@@ -16,11 +16,11 @@ import initializeContext from "./initializeContext";
...
  */
 export default async function invocationValidator(
-  executionContext: IntegrationExecutionContext<IntegrationInvocationEvent>
+  executionContext: IntegrationExecutionContext<IntegrationInvocationEvent>,
 ) {
   const wazuhExectionContext = initializeContext(executionContext);
 
-  wazuhExectionContext.provider.fetchManagerInfo().catch(error => {
+  wazuhExectionContext.provider.fetchManager().catch(error => {
     throw new Error(`${error}: validating provider connection`);
```
Comment (aiwilliams) [inline src/invocationValidator.ts]: This should throw the SDK-provided type `IntegrationInstanceAuthenticationError` when it is an authentication error, and `IntegrationInstanceConfigError` when there are missing values or the config is otherwise known to be mis-configured. Those types are handled in the runtime in such a way to inform the user through the job log UI. Other runtime errors will send a Pager Duty notification to on-call folks.

Check out https://github.com/JupiterOne/jupiter-integration-google/blob/master/src/invocationValidator.ts.

File: package.json (hunk 26 above)
Comment (lincolnfzr) [reply package.json]: Yes, saw some good stuff.

File: src/provider.ts [hunk 32]
```diff
@@ -36,81 +36,68 @@ export interface Agent {
...
-      _providerConfig.userId === undefined ||
-      _providerConfig.userId === undefined ||
-      _providerConfig.userId === "" ||
-      _providerConfig.userId === ""
+      providerConfig.userId === undefined ||
+      providerConfig.userId === undefined ||
+      providerConfig.userId === "" ||
+      providerConfig.userId === ""
     ) {
       throw new Error("User ID and Password required");
```
Comment (aiwilliams) [inline src/provider.ts]: I think this code would be good in the `invocationValidator`, to ensure that the config provided by the user is correct and throw the `IntegrationInstanceConfigError` from the `invocationValidator`. Then, the provider does not need to validate the config.

File: src/converters.test.ts (hunk 27 above)
Comment (lincolnfzr) [reply src/converters.test.ts]: got it

File: src/converters.test.ts (hunk 28 above)
Comment (lincolnfzr) [reply src/converters.test.ts]: Got it.  I missed that point.  I flipped them.

File: src/executionHandler.ts (hunk 29 above)
Comment (lincolnfzr) [reply src/executionHandler.ts]: No reason.  I was adding a message.   I am throwing the the unwrapped error.  Is that appropriate?

Comment (aiwilliams) [reply src/executionHandler.ts]: You may drop the `try/catch` altogether and let runtime errors get handled by the caller of the execution handler.

File: src/invocationValidator.ts (hunk 31 above)
Comment (lincolnfzr) [reply src/invocationValidator.ts]: Updated provider, invocationValidator, initializeContext to throw the appropriate error.  Updated test cases.

File: src/provider.ts (hunk 32 above)
Comment (lincolnfzr) [reply src/provider.ts]: I did remove the code from the provider and placed the responsibility on the initializeContext function to validate the config before returning the provider. `invocationValidator` makes a call to initializeContext then makes an API call to Wazuh through the provider client.

File: src/initializeContext.test.ts (hunk 30 above)
Comment (lincolnfzr) [reply src/initializeContext.test.ts]: That's correct.  This would allow tests against a test Wazuh API server.  It could be set before each stage [dev/int/performance/acceptance/] of the ci/cd pipeline. 

I will update the assignment statements.
